### PR TITLE
test(NODE-4965): unskip FLE range index spec tests and implement spec prose tests

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

This PR adds spec tests for Decimal128 range index queries, unskips the existing range index tests and adds all the prose tests related to range indexes.

CI will be mostly red until https://github.com/mongodb/libmongocrypt/pull/533 merges.  Serverless and mongosh failures are failing on main too.

We can confirm that the changes work as expected on the custom dependency tests from a pinned commit - logs [here](https://parsley.mongodb.com/evergreen/mongo_node_driver_next_rhel8_custom_dependency_tests_run_custom_csfle_tests_latest_pinned_commit_patch_056c86f6dee36f805376513c8322ac1ecab3c8e6_63cd464b2a60ed4384ce8f78_23_01_22_14_21_00/0/task?bookmarks=0,2253&selectedLine=1907).

##### Is there new documentation needed for these changes?

Not in the driver, no.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts [hunk 1]
```diff
@@ -0,0 +1,542 @@
+import { EJSON } from 'bson';
```
Comment (dariakp) [inline test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts:1]: I took a quick look at our file structure and we do have an established convention for the naming of these tests (see `client_side_encryption.prose.corpus.test.ts`), so let's follow it here, too; I suggest `client_side_encryption.prose.range.test.ts` but we can say range_explicit or whatever feels right to you

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts [hunk 2]
```diff
@@ -32,21 +32,21 @@ function byId(a, b) {
...
-    );
+    return (value, expected) => {
+      expect(value).to.be.instanceof(Decimal128);
+      expect((value as Decimal128).bytes).to.deep.equal(new Decimal128(expected.toString()).bytes);
+    };
   } else {
-    expect(value).to.equal(expectedValue);
+    return (value, expected) => expect(value).to.equal(expected);
   }
 }
```
Comment (nbbeeken) [inline test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts:51]: Can we use chai's expect().deep.equal instead, in all cases, so we shouldn't need a helper? unless I am missing something

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+
+        const autoEncryptionOptions = {
+          keyVaultNamespace: 'keyvault.datakeys',
+          kmsProviders: getKmsProviders(),
+          bypassQueryAnalysis: true
+        };
+
+        encryptedClient = await this.configuration
+          .newClient({}, { autoEncryption: autoEncryptionOptions })
+          .connect();
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts:217]: bump

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts (hunk 2 above)
Comment (baileympearson) [reply test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts:51]: see my response to your other comment

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts (hunk 1 above)
Comment (baileympearson) [reply test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts:1]: My preference would be to just break it into a folder and then use a descriptive name.

`tests/integration/client-side-encryption/prose/range_explicit_encryptipnprose.test.ts` (or even `range_explicit_encryption.test.ts`) feels more logical than `client_side_encryption.prose.range.test.ts`.  It's also easier to determine from looking at the filename, when using a CMD+P for example.

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+/**
+ * a comparitor function to sort two documents by their _id
+ */
+function byId(a, b) {
+  if (a._id > b._id) return 1;
+  if (a._id < b._id) return -1;
+  return 0;
+}
+
+function compareNumericValues(
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts]: Sure let's keep compareNumeric then, and can we use deep.eq unless the type is double/long, just so we're only handling the exceptions specially

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+          const resultOrError = await clientEncryption
+            .encrypt(factory(201), {
+              keyId,
+              algorithm: 'RangePreview',
+              contentionFactor: 0,
+              rangeOptions
+            })
+            .catch(e => e);
+
+          expect(resultOrError).to.be.instanceOf(Error);
```
Comment (nbbeeken) [inline test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts:467]: We either want to check MongoCryptError here, or a portion of the error message, this is broad enough to accept any kind of error and could be overly permissive

**Review by nbbeeken — CHANGES_REQUESTED** [review]

**Review by nbbeeken — APPROVED** [review]

File: test/integration/client-side-encryption/client_side_encryption.range.prose.ts [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+          }
+        ) as unknown as Document;
+
+        // confirm that this is an ObjectId
+        keyId = keyDocument1._id;
+        const encryptedFields = EJSON.parse(await readEncryptedFieldsFile(dataType), {
+          relaxed: false
+        }) as unknown as Document;
+
+        const utilClient: MongoClient = await this.configuration.newClient().connect();
```
Comment (nbbeeken) [inline test/integration/client-side-encryption/client_side_encryption.range.prose.ts:158]: missed one connect usage here

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: test/integration/client-side-encryption/client_side_encryption.range.prose.ts (hunk 6 above)
Comment (baileympearson) [reply test/integration/client-side-encryption/client_side_encryption.range.prose.ts:158]: intentional - util client that's closed in the before (not a part of tests).  do you still want me to remove it?

**Review by nbbeeken — APPROVED** [review]

File: test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts (hunk 1 above)
Comment (nbbeeken) [reply test/integration/client-side-encryption/range_explicit_encryption.prose.test.ts]: I think we still want to keep `.test.ts` as the file ending if it has mocha tests in it (unlike the deadlock tests which export them)

**Review by nbbeeken — APPROVED** [review]
LGTM

**Review by dariakp — APPROVED** [review]
LGTM on the naming convention

**Review by nbbeeken — APPROVED** [review]

**Pull request merged**